Qute Asynchronous Data Resolution not working as stated in example

**Describe the bug**
I am trying to use asynchronous data resolution within qute data as mentioned here https://quarkus.io/blog/qute/#asynchronous-data-resolution-on-the-way-to-reactive.

But it is not working for me at all...

I am trying to setup a completion stage from panache reactive mongodb and it always gives me the same error...

```
Caused by: java.lang.IllegalStateException: Unsupported iterable type found in [produkt.name]
	- matching type: java.util.concurrent.CompletionStage<java.util.List<de.blume2000.produkt.adapter.active.web.ProduktDTO>> 
	- found in template [produktListPage.html] on line 11
```

Here is the code:

```kotlin
@Path("/reactiveProdukte")
@Produces(MediaType.TEXT_HTML)
class ReactiveProduktHtmlResource(private val reactiveProduktApplicationService: ReactiveProduktApplicationService) {

  @GET
  fun find(): Uni<String> {
    val produkte = reactiveProduktApplicationService.execute(ZeigeAlleProdukteCommand)
      .map(Produkt::toDTO)
      .collectItems()
      .asList()
      .subscribeAsCompletionStage()
    val produktListPage = ReactiveProduktHtmlTemplates.produktListPage(produkte).renderAsync()
    return Uni.createFrom().completionStage(produktListPage)
  }
}

@CheckedTemplate
object ReactiveProduktHtmlTemplates {

  @JvmStatic
  external fun produktListPage(produkte: CompletionStage<List<ProduktDTO>>): TemplateInstance
}
```

And the HTML:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Produktübersicht</title>
</head>
<body>
<h1>Produktübersicht</h1>
{#for produkt in produkte}
<li>
  <b>{produkt.name}</b> {produkt.beschreibung}
</li>
{/for}
</body>
</html>
```

Here is a dependency Tree: 
[deps.txt](https://github.com/quarkusio/quarkus/files/5422089/deps.txt)

What am I doing wrong?
Does this feature already exist?

There has been a zulip exchange regarding this issue: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/qute-asynchronous-data-resolution-example

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/Names.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/Names.java
@@ -4,6 +4,7 @@ import java.util.Collection;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.concurrent.CompletionStage;
 import java.util.stream.Stream;
 
 import org.jboss.jandex.DotName;
@@ -33,6 +34,7 @@ final class Names {
     static final DotName COLLECTION = DotName.createSimple(Collection.class.getName());
     static final DotName CHECKED_TEMPLATE = DotName.createSimple(CheckedTemplate.class.getName());
     static final DotName TEMPLATE_INSTANCE = DotName.createSimple(TemplateInstance.class.getName());
+    static final DotName COMPLETION_STAGE = DotName.createSimple(CompletionStage.class.getName());
 
     private Names() {
     }

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
@@ -1048,6 +1048,19 @@ public class QuteProcessor {
                 matchType = extractMatchType(closure, Names.ITERATOR, firstParamType);
             }
         }
+
+        // Handle CompletionStage specially
+        if (matchType == null && ValueResolverGenerator.hasCompletionStageInTypeClosure(match.clazz, index)) {
+            Set<Type> closure = Types.getTypeClosure(match.clazz, Types.buildResolvedMap(
+                    match.getParameterizedTypeArguments(), match.getTypeParameters(), new HashMap<>(), index), index);
+            Function<Type, Type> firstParamType = t -> t.asParameterizedType().arguments().get(0);
+            // CompletionStage<List<Item>> => List<Item>
+            match.type = extractMatchType(closure, Names.COMPLETION_STAGE, firstParamType);
+            match.clazz = index.getClassByName(match.type.name());
+            processLoopHint(match, index, expression);
+            return;
+        }
+
         if (matchType != null) {
             match.type = matchType;
             match.clazz = index.getClassByName(match.type.name());

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/asyncdata/AsyncDataTest.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/asyncdata/AsyncDataTest.java
@@ -1,0 +1,48 @@
+package io.quarkus.qute.deployment.asyncdata;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.qute.Template;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class AsyncDataTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(Client.class)
+                    .addAsResource(new StringAsset("{@io.quarkus.qute.deployment.asyncdata.Client client}"
+                            + "{#for token in client.tokens}"
+                            + "{token}:{/for}"), "templates/test1.html")
+                    .addAsResource(new StringAsset("{@io.quarkus.qute.deployment.asyncdata.Client client}"
+                            + "{#for token in tokens}"
+                            + "{token}:{/for}"), "templates/test2.html")
+                    .addAsResource(new StringAsset("{token}"), "templates/test3.html"));
+
+    @Inject
+    Template test1;
+
+    @Inject
+    Template test2;
+
+    @Inject
+    Template test3;
+
+    @Test
+    public void testAsyncData() {
+        assertEquals("alpha:bravo:delta:", test1.data("client", new Client()).render());
+        assertEquals("alpha:bravo:delta:", test2.data("tokens", new Client().getTokens()).render());
+        assertEquals("alpha", test3.data("token", CompletableFuture.completedFuture("alpha")).render());
+    }
+
+}

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/asyncdata/Client.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/asyncdata/Client.java
@@ -1,0 +1,16 @@
+package io.quarkus.qute.deployment.asyncdata;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class Client {
+
+    public CompletionStage<List<String>> getTokens() {
+        CompletableFuture<List<String>> tokens = new CompletableFuture<>();
+        tokens.complete(Arrays.asList("alpha", "bravo", "delta"));
+        return tokens;
+    }
+
+}

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/EvaluatorImpl.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/EvaluatorImpl.java
@@ -80,6 +80,7 @@ class EvaluatorImpl implements Evaluator {
         }
     }
 
+    @SuppressWarnings("unchecked")
     private CompletionStage<Object> resolve(EvalContextImpl evalContext, Iterator<ValueResolver> resolvers,
             boolean tryCachedResolver) {
 
@@ -91,6 +92,10 @@ class EvaluatorImpl implements Evaluator {
                     if (Result.NOT_FOUND.equals(r)) {
                         return resolve(evalContext, resolvers, false);
                     } else {
+                        if (r instanceof CompletionStage) {
+                            // If the result is a completion stage return it as is
+                            return (CompletionStage<Object>) r;
+                        }
                         return CompletableFuture.completedFuture(r);
                     }
                 });
@@ -118,6 +123,10 @@ class EvaluatorImpl implements Evaluator {
                 } else {
                     // Cache the first resolver where a result is found
                     ((PartImpl) evalContext.part).setCachedResolver(resolver);
+                    if (r instanceof CompletionStage) {
+                        // If the result is a completion stage return it as is
+                        return (CompletionStage<Object>) r;
+                    }
                     return CompletableFuture.completedFuture(r);
                 }
             });

--- a/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
+++ b/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
@@ -888,7 +888,7 @@ public class ValueResolverGenerator {
         }
     }
 
-    static boolean hasCompletionStageInTypeClosure(ClassInfo classInfo,
+    public static boolean hasCompletionStageInTypeClosure(ClassInfo classInfo,
             IndexView index) {
 
         if (classInfo == null) {

